Yank/multistate/multistatesampler.py

```diff
@@ -1237,11 +1242,26 @@ def _minimize_replica(self, replica_id, tolerance, max_iterations):
             replica_id + 1, self.n_replicas, initial_energy))
 
         # Minimize energy.
-        openmm.LocalEnergyMinimizer.minimize(context, tolerance, max_iterations)
+        if (max_iterations is None) or (max_iterations == 0):
+            fire_iterations = 1000
+        else:
+            fire_iterations = max_iterations
+        logger.debug('Using FIRE: tolerance {} max_iterations {}'.format(tolerance, fire_iterations))
+        try:
+            integrator.step(max_iterations)
+        except Exception as e:
+            if str(e) == 'Particle coordinate is nan':
+                logger.debug('NaN encountered in FIRE minimizer; falling back to L-BFGS after resetting positions')
+                sampler_state.apply_to_context(context)
+                openmm.LocalEnergyMinimizer.minimize(context, tolerance, max_iterations)
+            else:
+                raise e
 
         # Get the minimized positions.
         sampler_state.update_from_context(context)
 
+        # TODO: Do we have to destroy the FIRE integrator-associated context if we aren't going to use it again?
```

Comment:
Nope! Deleting the integrator has no practical effect if you are using the `ContextCache`. The context and the integrator will be purged automatically after a while.

Reply:
Will `ContextCache` reset the global parameters to the defaults in the requested `integrator` object before returning the context/integrator pair if it finds it already exists?